posts/ilaborde/01.automapper.md

```diff
@@ -0,0 +1,170 @@
+
+```c#
+    public class Blog
+    {
+        public int Id { get; set; }
+        public string Name { get; set; }
+        public ICollection<Post> Posts { get; set; }
+    }
+
+    public class Post
+    {
+        public int Id { get; set; }
+        public string Title { get; set; }
+        public Blog Blog { get; set; }
+        public List<Comment> Comments { get; set; }
+    }
+
+    public class Comment
+    {
+        public int Id { get; set; }
+        public string Content { get; set; }
+        public Post Post { get; set; }
+    }
+```
+
+And DTOs
+
+```c#
+    public class PostDto
+    {
+        public int Id { get; set; }
+        public string Title { get; set; }
+        public int BlogId { get; set; }
+        public string BlogName { get; set; }
+        public List<CommentDto> CommentsDto { get; set; }
+    }
+
+    public class CommentDto
+    {
+        public int Id { get; set; }
+        public string Content { get; set; }
+    }
+
+    public class PostSummaryDto
+    {
+        public int Id { get; set; }
+        public string Title { get; set; }
+        public int CommentsCount { get; set; }
+    }
+```
+
+When querying data with Entity Framework is very common to query data using the Include method, and then manually map the collection of posts to a collection of `PostDto`:
+
+```c#
+  var posts = dbContext.Posts
+      .Include(x => x.Blog)
+      .Include(x => x.Comments)
+      .ToListAsync();
+
+ // manually map to DTO     
+```
+
+Entity Framework will generate the following query:
+
+```SQL
+    SELECT [p].[Id], [p].[BlogId], [p].[Title], [b].[Id], [b].[Name], 
+    [c].[Id], [c].[Content], [c].[PostId]
+    FROM [Posts] AS [p]
+    INNER JOIN [Blogs] AS [b] ON [p].[BlogId] = [b].[Id]
+    LEFT JOIN [Comments] AS [c] ON [p].[Id] = [c].[PostId]
+    ORDER BY [p].[Id], [b].[Id]
+```
+
+Or if you use AutoMapper, you'll probably do something like this instead:
+
+```c#
+  var posts = dbContext.Posts
+      .Include(x => x.Blog)
+      .Include(x => x.Comments)
+      .ToListAsync();
+  var collectionOfPostDto = mapper.Map<PostDto>(posts);
+```
+
+Obviously generating the same SQL query.
+
+You can do the same using an Entity Framework projection, just picking the fields that you need:
+
+```c#
+    var post = dbContext.Posts
+        .Select(x => new PostDto()
+        {
+            Id = x.Id,
+            Title = x.Title,
+            BlogId = x.Blog.Id,
+            BlogName = x.Blog.Name,
+            CommentsDto = x.Comments.Select(c => new CommentDto()
+            {
+                Id = c.Id,
+            }).ToList()
+        })
+        .ToListAsync();
+```
+
+With the following query as the result:
+
+```SQL
+    SELECT [p].[Id], [p].[Title], [b].[Id], [b].[Name], [c].[Id]
+    FROM [Posts] AS [p]
+    INNER JOIN [Blogs] AS [b] ON [p].[BlogId] = [b].[Id]
+    LEFT JOIN [Comments] AS [c] ON [p].[Id] = [c].[PostId]
+    ORDER BY [p].[Id], [b].[Id]
+```
+
+### ProjectTO to the rescue
```

Comment:
It says "to the rescue" of some problem, but the problem was never actually mentioned. For example, in the first query you show, you probably want to mention that full entities are meterialized, including properties you might not even need.
The second alternative with a manual projection would solve that, but what problems does it have using it that way? THAT is what projecTo solves and justifies the "to the rescue".